trollmoves/client.py

```diff
@@ -728,25 +734,50 @@ def config_equals(self, other_config):
                 return False
         return True
 
+    def get_unchanged_providers(self, other_config):
+        """Get a list of providers that have not changed between this and other config."""
+        if self._config["topic"] != other_config["topic"]:
+            return []
+        return list(set(self._config["providers"]).intersection(set(other_config["providers"])))
+
     def publisher_needs_restarting(self, other_config):
         """Check that current config is the same as `other_config`."""
         for key in ["nameservers", "publish_port"]:
             if self._config[key] != other_config[key]:
                 return True
         return False
 
-    def reset_listeners(self):
+    def refresh(self, new_config, callback):
+        """Refresh the chain with new config."""
+        publisher_needs_restarting = self.publisher_needs_restarting(new_config)
+        unchanged_providers = self.get_unchanged_providers(new_config)
+        self._config = new_config
+        self.stop(stop_publisher=publisher_needs_restarting, keep_providers=unchanged_providers)
+        self.setup_publisher()
+        self.setup_listeners(callback, keep_providers=unchanged_providers)
+        if not self.running:
+            self.start()
```

Comment:
This could be split into refreshing the publishers and refreshing the listeners for clarity. Maybe to that effect, the `stop` method should be split into `stop_listenerers` and `stop_publishers` ?

Reply:
Refactored the refreshing. With that, the `stop` method also got much cleaner.